test/WebSites/InProcessWebSite/Startup.cs

```diff
@@ -424,6 +445,110 @@ private void WebsocketRequest(IApplicationBuilder app)
             });
         }
 
+
+        private void WebSocketEcho(IApplicationBuilder app)
+        {
+            app.Run(async context =>
+            {
+                var ws = await Upgrade(context);
+
+                var appLifetime = app.ApplicationServices.GetRequiredService<IApplicationLifetime>();
+
+                await Echo(ws, appLifetime.ApplicationStopping);
+            });
+        }
+
+        private void WebSocketLifetimeEvents(IApplicationBuilder app)
+        {
+            app.Run(async context => {
+
+                var messages = new List<string>();
+
+                context.Response.OnStarting(() => {
+                    context.Response.Headers["custom-header"] = "value";
+                    messages.Add("OnStarting");
+                    return Task.CompletedTask;
+                });
+
+                var ws = await Upgrade(context);
+                messages.Add("Upgraded");
+
+                await SendMessages(ws, messages.ToArray());
+            });
+        }
+
+        private static async Task SendMessages(WebSocket webSocket, params string[] messages)
+        {
+            foreach (var message in messages)
+            {
+                await webSocket.SendAsync(new ArraySegment<byte>(Encoding.ASCII.GetBytes(message)), WebSocketMessageType.Text, true, CancellationToken.None);
+            }
+        }
+
+        private static async Task<WebSocket> Upgrade(HttpContext context)
```

Comment:
Move websocket stuff into a separate class, such that it can be used by other tests in the future.